pages/studentAssessmentExam/studentAssessmentExam.js

```diff
@@ -8,25 +8,46 @@ var sqldb = require('../../prairielib/lib/sql-db');
 var sqlLoader = require('../../prairielib/lib/sql-loader');
 
 var sql = sqlLoader.loadSqlEquiv(__filename);
+var groupAssessmentHelper = require('../../lib/groups');
 
 router.get('/', function (req, res, next) {
   if (res.locals.assessment.type !== 'Exam') return next();
+  var params = {
+    assessment_id: res.locals.assessment.id,
+    user_id: res.locals.user.user_id,
+  };
   if (res.locals.assessment.multiple_instance) {
-    res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+    if (res.locals.assessment.group_work) {
+      groupAssessmentHelper.getGroupInfo(res, function (err) {
+        if (ERR(err, next)) return;
+        res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+      });
+    } else {
+      res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+    }
   } else {
-    var params = {
-      assessment_id: res.locals.assessment.id,
-      user_id: res.locals.user.user_id,
-    };
     sqldb.query(sql.select_single_assessment_instance, params, function (err, result) {
       if (ERR(err, next)) return;
       if (result.rowCount === 0) {
-        res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+        if (res.locals.assessment.group_work) {
+          groupAssessmentHelper.getGroupInfo(res, function (err) {
+            if (ERR(err, next)) return;
+            res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+          });
+        } else {
+          res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+        }
       } else {
         res.redirect(res.locals.urlPrefix + '/assessment_instance/' + result.rows[0].id);
       }
     });
   }
+  if (res.locals.assessment.group_work) {
+    groupAssessmentHelper.getGroupInfo(res, function (err) {
+      if (ERR(err, next)) return;
+      res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+    });
+  }
```

Comment:
I'm a bit surprised to see this here. I'd expect all of the preceding code to be sufficient, and I'd actually expect this bit to cause problems, since we'd probably end up calling `res.render` twice which is a big no-no.

I can test this for myself once merge conflicts are resolved!

Reply:
Yep, as expected, this produced a `ERR_HTTP_HEADERS_SENT` error:

```
Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at new NodeError (node:internal/errors:371:5)
    at ServerResponse.setHeader (node:_http_outgoing:576:11)
    at ServerResponse.header (/Users/nathan/git/PrairieLearn/node_modules/express/lib/response.js:776:10)
    at ServerResponse.send (/Users/nathan/git/PrairieLearn/node_modules/express/lib/response.js:170:12)
    at done (/Users/nathan/git/PrairieLearn/node_modules/express/lib/response.js:1013:10)
    at tryHandleCache (/Users/nathan/git/PrairieLearn/node_modules/ejs/lib/ejs.js:278:5)
    at View.exports.renderFile [as engine] (/Users/nathan/git/PrairieLearn/node_modules/ejs/lib/ejs.js:489:10)
    at View.render (/Users/nathan/git/PrairieLearn/node_modules/express/lib/view.js:135:8)
    at tryRender (/Users/nathan/git/PrairieLearn/node_modules/express/lib/application.js:640:10)
    at Function.render (/Users/nathan/git/PrairieLearn/node_modules/express/lib/application.js:592:3)
    at ServerResponse.render (/Users/nathan/git/PrairieLearn/node_modules/express/lib/response.js:1017:7)
    at module.exports (/Users/nathan/git/PrairieLearn/pages/error/error.js:64:9)
    at Layer.handle_error (/Users/nathan/git/PrairieLearn/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/Users/nathan/git/PrairieLearn/node_modules/express/lib/router/index.js:321:13)
    at /Users/nathan/git/PrairieLearn/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/nathan/git/PrairieLearn/node_modules/express/lib/router/index.js:341:12)
    at next (/Users/nathan/git/PrairieLearn/node_modules/express/lib/router/index.js:275:10)
    at module.exports (/Users/nathan/git/PrairieLearn/middlewares/redirectEffectiveAccessDenied.js:11:34)
    at Layer.handle_error (/Users/nathan/git/PrairieLearn/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/Users/nathan/git/PrairieLearn/node_modules/express/lib/router/index.js:321:13)
    at /Users/nathan/git/PrairieLearn/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/nathan/git/PrairieLearn/node_modules/express/lib/router/index.js:341:12)
```

Reply:
Ah yes, that was bad. Removed in [d1d3c69](https://github.com/PrairieLearn/PrairieLearn/pull/5290/commits/d1d3c69257b2e95a910abf49bdcf856ff8f58b6c)